samples/nfc/system_off/src/main.c

```diff
@@ -0,0 +1,195 @@
+				      sizeof(en_code),
+				      en_payload,
+				      sizeof(en_payload));
+
+	NFC_NDEF_MSG_DEF(nfc_text_msg, MAX_REC_COUNT);
+
+	u32_t len = sizeof(buffer);
+
+	/* Set up NFC */
+	if (nfc_t2t_setup(nfc_callback, NULL) < 0) {
+		printk("Cannot setup NFC T2T library!\n");
+		return -1;
+	}
+
+	/* Add record */
+	if (nfc_ndef_msg_record_add(&NFC_NDEF_MSG(nfc_text_msg),
+				&NFC_NDEF_TEXT_RECORD_DESC(nfc_text_rec)) < 0) {
+		printk("Cannot add record!\n");
+		return -1;
+	}
+
+	/* Encode welcome message */
+	if (nfc_ndef_msg_encode(&NFC_NDEF_MSG(nfc_text_msg),
+				buffer, &len) < 0) {
+		printk("Cannot encode message!\n");
+		return -1;
+	}
+
+	/* Set created message as the NFC payload */
+	if (nfc_t2t_payload_set(buffer, len) < 0) {
+		printk("Cannot set payload!\n");
+		return -1;
+	}
+
+	/* Start sensing NFC field */
+	if (nfc_t2t_emulation_start() < 0) {
+		printk("Cannot start emulation!\n");
+		return -1;
+	}
+
+	printk("NFC configuration done\n");
+	return 0;
+}
+
+
+/**
+ * @brief Function entering system off.
+ * System off is delayed to make sure that NFC tag was correctly read.
+ */
+static void system_off(struct k_work *work)
+{
+	printk("Entering system off.\nApproach a NFC reader to restart.\n");
+
+	/* Before we disabled entry to deep sleep. Here we need to override
+	 * that, then force a sleep so that the deep sleep takes effect.
+	 */
+	sys_pm_force_power_state(SYS_POWER_STATE_DEEP_SLEEP_1);
+
+	dk_set_led_off(SYSTEM_ON_LED);
+
+	/* Going into sleep will actually go to system off mode, because we
+	 * forced it above.
+	 */
+	k_sleep(K_MSEC(1));
+
+	/* k_sleep will never exit, so below two lines will never be executed
+	 * if system off was correct. On the other hand if someting gone wrong
+	 * we will see it on terminal and LED.
+	 */
+	dk_set_led_on(SYSTEM_ON_LED);
+	printk("ERROR: System off failed\n");
+}
+
+
+void main(void)
+{
+	u32_t reas;
+
+	/* Configure LED-pins */
+	if (dk_leds_init() < 0) {
+		printk("Cannot init LEDs!\n");
+		return;
+	}
+	dk_set_led_on(SYSTEM_ON_LED);
+
+	/* Configure and start delayed work that enters system off */
+	k_delayed_work_init(&system_off_work, system_off);
+	k_delayed_work_submit(&system_off_work, K_SECONDS(SYSTEM_OFF_DELAY_S));
+
+	/* Show last reset reason */
+	reas = nrf_power_resetreas_get(NRF_POWER);
+	nrf_power_resetreas_clear(NRF_POWER, reas);
+	if (reas & NRF_POWER_RESETREAS_NFC_MASK) {
+		printk("Wake up by NFC field detect\n");
+	} else if (reas & NRF_POWER_RESETREAS_RESETPIN_MASK) {
+		printk("Reset by pin-reset\n");
+	} else if (reas & NRF_POWER_RESETREAS_SREQ_MASK) {
+		printk("Reset by soft-reset\n");
+	} else if (reas) {
+		printk("Reset by a different source (0x%08X)\n", reas);
+	} else {
+		printk("Power-on-reset\n");
+	}
+
+	/* Start NFC */
+	if (start_nfc() < 0) {
+		printk("ERROR: NFC configuration failed\n");
+		return;
+	}
+
+	/* Prevent deep sleep (system off) from being entered */
+	sys_pm_ctrl_disable_state(SYS_POWER_STATE_DEEP_SLEEP_1);
```

Comment:
If this deep sleep was disabled here, the system would enter deep sleep on the exit of `main()`?

Reply:
It may, but in our specific case no. It will go if we are sleeping more than `CONFIG_SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_1` (which is 60 sec. by default) , but in our case delayed work will be executed after 3 sec. I added it to make sure that sample will work in any PM configuration.